daemon/fmutex/fmutex.go

```diff
@@ -0,0 +1,81 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2015 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package fmutex implements locking for the snappy daemon
+package fmutex
+
+import (
+	"sync"
+
+	"github.com/ubuntu-core/snappy/dirs"
+	"github.com/ubuntu-core/snappy/logger"
+	"github.com/ubuntu-core/snappy/priv"
+)
+
+// FMutex gives you an exclusive mutex that also holds the filesystem lock.
+type FMutex struct {
+	mutex sync.Locker
+	flock FLocker
+}
+
+var _ sync.Locker = (*FMutex)(nil)
+
+// FLocker is the part of priv.Mutex that we use.
+type FLocker interface {
+	Lock() error
+	Unlock() error
+}
+
+func flockImpl() FLocker {
+	return priv.New(dirs.SnapLockFile)
+}
+
+// NewFLock constructs a new FLocker and returns it.
+//
+// In the default implementation it's a priv.Mutex. Exposed for
+// testing, as priv.Mutex requires root.
+var NewFLock = flockImpl
```

Comment:
That would be even better. I don't see why a file-locking implementation depends on being root.